### ISSUE TITLE
DDoS-ing the Metadata Server.

### ISSUE THREAD

**Issue by xvzf** [googleapis/google-auth-library-php#297]
Hello,

With high volume traffic, a check (on each interaction with Google APIs!!!) if the application is running on GCP by calling the metadata server is becoming quite a bottleneck response-time wise.

The specific code is located here:
https://github.com/googleapis/google-auth-library-php/blob/c438cdaadd2b1fa334c4da7560cad238625d8c10/src/Credentials/GCECredentials.php#L258

#### Environment details
  - OS: Linux
  - PHP version: 7.x
  - Package name and version: irrelevant

**Comment by xvzf** [comment]
@dwsupplee since you tagged it as investigating, it's not just PHP 7.4 but all 7.x releases

**Comment by ericnorris** [comment]
At Etsy we've encountered the same issue; during periods of high enough traffic (e.g. backfills) we've seen metadata server outages that result in the following backtrace:

```
Could not load the default credentials. Browse to https://developers.google.com/accounts/docs/application-default-credentials for more information
DomainException trace:
[Dev_Backtrace]:
#12 getCredentials() /GoogleCloudCredentials.php:70
...
```

...where the `getCredentials` is this call:

 https://github.com/googleapis/google-auth-library-php/blob/c05c0a5c24759949f16c8d868858f05ab3899413/src/ApplicationDefaultCredentials.php#L148

...which ends up calling the `onGce` method in the original comment.

We use the `FetchAuthTokenCache` to avoid hammering the metadata server with auth calls, but this `onGce` call went unnoticed until we experience outages. Internally we are trying to resolve this by passing in a special `$httpHandler` to `ApplicationDefaultCredentials::getCredentials`, but ideally the library would offer a PSR-6 way of caching this, much like how the `FetchAuthTokenCache` works.

**Comment by bshaffer** [comment]
If we added a new class, like @ericnorris suggested, we could have something like this:

```php
// GCE cache implementing CacheItemPoolInterface
$cache = new FileCache();
// optional GCE cache config
$cacheConfig = [
    'lifetime' => 1500,
    'prefix' => '',
];
$gceCache = new GCECache($cache, $cacheConfig);

// fetch default credentials with the cache
$credentials = ApplicationDefaultCredentials::getCredentials($scope, null, null, null, null, $gceCache);
// check directly with the cache
$onGce = GCECredentials::getCredentials(null, $gceCache);
```

We can't use the existing `$cache` and `$cacheConfig` parameters passed to `ApplicationDefaultCredentials::getCredentials` because the cache config / cache interface may be different for the GCE metadata cache.

We could also add `CacheItemPoolInterface` and `$cacheConfig` as additional parameters and use `GCECache under the hood. This would be messier, but at least it would be consistent:

```php
// FetchAuthTokn cache implementing CacheItemPoolInterface
$authTokenCache = new FileCache();
// optional GCE cache config
$authTokenCacheConfig = [
    'lifetime' => 1500,
    'prefix' => '',
];

// GCE cache implementing CacheItemPoolInterface
$gceCache = new FileCache();
// optional GCE cache config
$gceCacheConfig = [
    'lifetime' => 1500,
    'prefix' => '',
];

// fetch default credentials with the cache
$credentials = ApplicationDefaultCredentials::getCredentials(
    $scope,
    null,
    null,
    $authTokenCache,
    $authTokenCacheConfig,
    $gceCache,
    $gceCacheConfig
);
// check directly with the cache
$onGce = GCECredentials::getCredentials(null, $gceCache, $gceCacheConfig);
```

Do these solutions sound sufficient, and do you have a preference for either of them?

**Comment by ericnorris** [comment]
@bshaffer, I actually proposed a different solution internally (props to @etsteeeve for implementing) that seems to work for us. We've instead written a wrapper class called `LazyApplicationDefaultCredentials` that only calls `ApplicationDefaultCredentials::getCredentials` when `::fetchAuthToken` and `::getLastReceivedToken` is called on it. Since this is passed into a `FetchAuthTokenCache`, this means the `::onGce` call is only made once(ish) every ~25 minutes or so.

Perhaps this could be implemented in the library as a `DefaultCredentialsFetcher`?

**Comment by bshaffer** [comment]
@ericnorris This is a really great idea! If you or @etsteeeve want to submit a PR please do! Otherwise I can take a look.

It would be nice to have caching for `::onGce` independently, so we could implement both of these options.

**Comment by bshaffer** [comment]
I've coded up a potential option, please tell me what you think.

**Comment by bshaffer** [comment]
fixed in `v1.12.0`

**Comment by Xfaider48** [comment]
@bshaffer I disagree with the provided soultion. 

I think [this code](https://github.com/googleapis/google-auth-library-php/blob/c438cdaadd2b1fa334c4da7560cad238625d8c10/src/Credentials/GCECredentials.php#L258-L290) must be disabled by default. I'm sure that this is a big problem when without any credentials you get enabled this mechanism and it ruins any feature tests. E.g. when you use datastore/spanner (or something else with GRPC) in feature tests and you don't need any credentials (because you can use datastore/spanner emulators) you get each test slowed down by at least 1.5 sec (3 tries with 500ms timeout per GRPC class). And [the default caching mechanism](https://github.com/googleapis/google-cloud-php/blob/main/Core/src/RequestWrapperTrait.php#L103) doesn't work because it's in-memory storage and a state of an app is cleared between test cases.
